libcni/api.go

```diff
@@ -223,17 +235,99 @@ func getCachedResult(netName, cniVersion string, rt *RuntimeConf) (types.Result,
 }
 
 // GetNetworkListCachedResult returns the cached Result of the previous
-// previous AddNetworkList() operation for a network list, or an error.
+// AddNetworkList() operation for a network list, or an error.
 func (c *CNIConfig) GetNetworkListCachedResult(list *NetworkConfigList, rt *RuntimeConf) (types.Result, error) {
 	return getCachedResult(list.Name, list.CNIVersion, rt)
 }
 
 // GetNetworkCachedResult returns the cached Result of the previous
-// previous AddNetwork() operation for a network, or an error.
+// AddNetwork() operation for a network, or an error.
 func (c *CNIConfig) GetNetworkCachedResult(net *NetworkConfig, rt *RuntimeConf) (types.Result, error) {
 	return getCachedResult(net.Network.Name, net.Network.CNIVersion, rt)
 }
 
+type cniCachedFile struct {
+	Kind         string       `json:"kind"`
+	CachedConfig cachedConfig `json:"cachedConfig"`
+}
+
+type cachedConfig struct {
+	CniVersion     string                 `json:"cniVersion"`
+	Config         []byte                 `json:"config"`
+	CniArgs        string                 `json:"cniArgs,omitempty"`
+	CapabilityArgs map[string]interface{} `json:"capabilityArgs,omitempty"`
+}
+
+func setCachedConfig(config []byte, netName string, rt *RuntimeConf) error {
+	fname := getCacheFilePath(cacheFileTypeConfig, netName, rt)
+	if err := os.MkdirAll(filepath.Dir(fname), 0700); err != nil {
+		return err
+	}
+
+	cached := cachedConfig{Config: config,
+		CniVersion: "0.4.0"}
+	if len(rt.Args) > 0 {
+		cached.CniArgs = utils.StringifyArgs(rt.Args)
+	}
+	if len(rt.CapabilityArgs) > 0 {
+		cached.CapabilityArgs = rt.CapabilityArgs
+	}
+	kind := cniCachedFile{Kind: "cniCacheV1",
+		CachedConfig: cached}
+
+	newBytes, err := json.Marshal(&kind)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(fname, newBytes, 0600)
+}
+
+func delCachedConfig(netName string, rt *RuntimeConf) error {
+	fname := getCacheFilePath(cacheFileTypeConfig, netName, rt)
+	return os.Remove(fname)
+}
+
+func getCachedConfig(netName string, rt *RuntimeConf) ([]byte, *RuntimeConf, error) {
+	fname := getCacheFilePath(cacheFileTypeConfig, netName, rt)
+	bytes, err := ioutil.ReadFile(fname)
+	if err != nil {
+		// Ignore read errors; the cached result may not exist on-disk
+		return nil, nil, nil
+	}
+
+	unmarshaled := cniCachedFile{}
+	if err := json.Unmarshal(bytes, &unmarshaled); err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal cached network %q config: %v", netName, err)
+	}
+	if unmarshaled.Kind != "cniCacheV1" {
+		return nil, nil, fmt.Errorf("read cached network %q config has wrong kind: %v", netName, unmarshaled.Kind)
+	}
+	cached := unmarshaled.CachedConfig
+
+	newRt := &RuntimeConf{}
+	if cached.CniArgs != "" {
+		if newRt.Args, err = utils.ParseArgs(cached.CniArgs); err != nil {
+			return nil, nil, fmt.Errorf("failed to parse cached config CNI Args %q: %v", cached.CniArgs, err)
+		}
+	}
+	newRt.CapabilityArgs = cached.CapabilityArgs
```

Comment:
I think **newRt** here should be filled with ContainerID, NetNS and so on, IfName, because we may use it to DelNetworkList directly and these fields are necessary.

Reply:
Yeah, seems a better way ~